### PR TITLE
add textual diffing tab based on diffobj, addresses #44

### PR DESCRIPTION
**Pull request by cpsievert**
For the diffing "widget" to span the full width, you'll need `devtools::install_github("brodieG/diffobj@development")`

### CODE REVIEW

**Comment by lionel-** [comment]
Was the write_svg change necessary to this PR?

File: R/widgets.R [hunk 1]
```diff
@@ -68,23 +68,39 @@ widget_diff_ <- function(before, after, width = NULL, height = NULL) {
...
+        style = list(html.output = 'diff.w.style')
+      )
+    )
+  )
+}
+
 #' @rdname htmlwidgets
 #' @export
-widget_toggle <- function(before, after, width = NULL, height = NULL) {
+widget_toggle <- function(before, after, input, width = NULL, height = NULL) {
```
Comment (lionel-) [inline R/widgets.R]: This new argument needs a doc bullet. Should it have a default so it doesn't look like a compulsory argument?

File: R/shiny-server.R [hunk 2]
```diff
@@ -89,13 +110,14 @@ renderDiffer <- function(input, active_cases, widget) {
...
 
-    widget(before, after)
+    widget(before, after, input)
   })
 }
 
 withdraw_cases <- function(cases) {
   validate_cases(filter_cases(cases, c("new_case", "mismatch_case")))
-  delete_orphaned_cases(filter_cases(cases, "orphaned_case"))
+  delete_orphaned_cases(filter_cases
```
Comment (lionel-) [inline R/shiny-server.R]: whitespace?

**Comment by cpsievert** [comment]
> Was the write_svg change necessary to this PR?

No, I could revert that commit if you like, but it'd be great to have that PR merged as well.

> Would it be possible for the diff to take up all horizontal space? It looks quite narrow:

This is due to a bug in diffobj which has been fixed in  `devtools::install_github("brodieG/diffobj@development")`

File: R/shiny-server.R (hunk 2 above)
Comment (cpsievert) [reply R/shiny-server.R]: whoops, yea, I'll fix this

**Comment by lionel-** [comment]
> No, I could revert that commit if you like, but it'd be great to have that PR merged as well.

Yes, PRs should be independent if possible. It facilitates reviews and doesn't tie their acceptance.

File: R/widgets.R (hunk 1 above)
Comment (cpsievert) [reply R/widgets.R]: Yea, `input = NULL` would be better

**Comment by cpsievert** [comment]
I guess I decided to branch off the write-svg branch so I could install from one branch to run my visual tests. Happy to revert if and when these changes are approved.

**Comment by lionel-** [comment]
@cpsievert Can you please take a look at the new approach I just pushed? Now the widgets take `...` for additional arguments. The difftext widget takes an actual mode parameter rather than input. The reading of input is done inside `renderDiffer()` which takes a watcher function to read `input` and create additional arguments passed to widgets.

This seems a better design in terms of extension and separation of concerns. Should we export the difftext widget?

**Comment by cpsievert** [comment]
Looks great, thanks!!

> Should we export the difftext widget?

I have no string feelings either way -- I personally don't yet have a use case for directly using the widgets outside of `manage_cases()`. If we do export, we should wrap the result in `htmltools::browsable()` so the HTML is actually rendered when the output is printed.

**Comment by lionel-** [comment]
Ok let's wait and see if a need to export arises later.

**Comment by lionel-** [comment]
Thanks, that's a great addition! And thanks to @brodieG for diffObj.

**Pull request merged**